storage-manager/src/S3Storage.cpp

```diff
@@ -0,0 +1,514 @@
+}
+
+// convenience macro for testConnectivityAndPerms()
+#define FAIL(OP) { \
+        const char *msg = "S3Storage: failed to " #OP ", check log files for specific error"; \
+        logger->log(LOG_ERR, msg); \
+        throw runtime_error(msg); \
+}
+
+void S3Storage::testConnectivityAndPerms()
+{
+    boost::shared_array<uint8_t> testObj(new uint8_t[1]);
+    testObj[0] = 0;
+    string testObjKey("connectivity_test");
+    
+    int err = putObject(testObj, 1, testObjKey);
+    if (err)
+        FAIL(PUT)
+    bool _exists;
+    err = exists(testObjKey, &_exists);
+    if (err)
+        FAIL(HEAD)
+    size_t len;
+    err = getObject(testObjKey, &testObj, &len);
+    if (err)
+        FAIL(GET)
+    err = deleteObject(testObjKey);
+    if (err)
+        FAIL(DELETE)
+    logger->log(LOG_DEBUG, "S3Storage: S3 connectivity & permissions are OK");
+}
+
+int S3Storage::getObject(const string &sourceKey, const string &destFile, size_t *size)
+{
+    int fd, err;
+    boost::shared_array<uint8_t> data;
+    size_t len, count = 0;
+    char buf[80];
+    
+    err = getObject(sourceKey, &data, &len);
+    if (err)
+        return err;
+    
+    fd = ::open(destFile.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0600);
+    if (fd < 0)
+    {
+        int l_errno = errno;
+        logger->log(LOG_ERR, "S3Storage::getObject(): Failed to open %s, got %s", destFile.c_str(), 
+            strerror_r(l_errno, buf, 80));
+        errno = l_errno;
+        return err;
+    }
+    ScopedCloser s(fd);
+    while (count < len)
+    {
+        err = ::write(fd, &data[count], len - count);
+        if (err < 0)
+        {
+            int l_errno = errno;
+            logger->log(LOG_ERR, "S3Storage::getObject(): Failed to write to %s, got %s", destFile.c_str(), 
+              strerror_r(errno, buf, 80));
+            errno = l_errno;
+            return -1;
+        }
+        count += err;
+    }
+    if (size)
+        *size = len;
+    return 0;
+}
+
+int S3Storage::getObject(const string &_sourceKey, boost::shared_array<uint8_t> *data, size_t *size)
+{
+    uint8_t err;
+    size_t len = 0;
+    uint8_t *_data = NULL;
+    string sourceKey = prefix + _sourceKey;
+    
+    ms3_st *creds = getConnection();
+    ScopedConnection sc(this, creds);
+    
+    do {
+        err = ms3_get(creds, bucket.c_str(), sourceKey.c_str(), &_data, &len);
+        if (err && retryable_error(err))
+        { 
+            if (ms3_server_error(creds))
+                logger->log(LOG_WARNING, "S3Storage::getObject(): failed to GET, server says '%s'.  bucket = %s, key = %s."
+                    "  Retrying...", ms3_server_error(creds), bucket.c_str(), sourceKey.c_str());
+            else 
+                logger->log(LOG_WARNING, "S3Storage::getObject(): failed to GET, got '%s'.  bucket = %s, key = %s.  Retrying...", 
+                    s3err_msgs[err], bucket.c_str(), sourceKey.c_str());
+            sleep(5);
+        }
+    } while (err && retryable_error(err));
```

Comment:
If S3 is down or there are WAN issues, won't this become an infinite loop? Maybe abort after X tries?